File: AULA 2/C#/D - Parentheses Balance.cs
using System;
using System.Linq;
using System.Collections.Generic;

public static class Program
{
	public static void Main()
	{
		var result = "";
		
		var simbolDict = new Dictionary<char, char>
		{
			{ '[', ']' },
			{ '(', ')' }
		};
		
		var numCases = int.Parse(Console.ReadLine());
		
		for(var i = 0; i < numCases; i++)
		{
			var seqQueue = new List<char>();
			var sequenceStr = Console.ReadLine();
			var sequence = Console.ReadLine().ToCharArray();
			
			if(string.IsNullOrEmpty(sequenceStr))
			{
				result += "Yes\n";
				continue;
			}
			
			var isRight = true;
			foreach(var simbol in sequence)
			{
				if(simbol == '(' || simbol == '[')
				{
					seqQueue.push(simbol);
				}
				else if ((simbol == ')' || simbol == ']') && (seqQueue.Count == 0 || simbolDict[seqQueue.pop()] != simbol))
				{
					isRight = false;
					break;
				}
			}
			
			result += isRight ? "Yes\n" : "No\n"; 
		}
		
		Console.Write(result);
	}
	
	public static char pop(this List<char> list) { var item = list.ElementAt(list.Count - 1); list.RemoveAt(list.Count - 1); return item; }
	public static void push(this List<char> list, char item) { list.Add(item); }
}

File: AULA 2/Java/D - Parentheses Balance.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        StringBuilder result = new StringBuilder();
        
        Map<Character, Character> simbolDict = new HashMap<>();
        simbolDict.put('[', ']');
        simbolDict.put('(', ')');
        
        int numCases = Integer.parseInt(scanner.nextLine());
        
        for (int i = 0; i < numCases; i++) {
            List<Character> seqQueue = new ArrayList<>();
            String sequenceStr = scanner.nextLine();
            char[] sequence = sequenceStr.toCharArray();
            
            if (sequenceStr.isEmpty()) {
                result.append("Yes\n");
                continue;
            }
            
            boolean isRight = true;
            for (char simbol : sequence) {
                if (simbol == '(' || simbol == '[') {
                    seqQueue.add(simbol);
                } else if ((simbol == ')' || simbol == ']') && (seqQueue.isEmpty() || simbolDict.get(seqQueue.remove(seqQueue.size() - 1)) != simbol)) {
                    isRight = false;
                    break;
                }
            }
            
            result.append(isRight && seqQueue.isEmpty() ? "Yes\n" : "No\n");
        }
        
        System.out.print(result.toString());
        scanner.close();
    }
}